src/cargo/core/package.rs

```diff
@@ -571,6 +558,80 @@ impl<'cfg> PackageSet<'cfg> {
         Ok(())
     }
 
+    /// Check if there are any dependency packages that do not have any libs.
+    pub(crate) fn no_lib_pkgs(
+        &self,
+        resolve: &Resolve,
+        root_ids: &[PackageId],
+        has_dev_units: HasDevUnits,
+        requested_kinds: &[CompileKind],
+        target_data: &RustcTargetData<'_>,
+        force_all_targets: ForceAllTargets,
+    ) -> BTreeMap<PackageId, Vec<&Package>> {
+        let mut ret = BTreeMap::new();
+
+        root_ids.iter().for_each(|&root_id| {
+            let pkgs: Vec<&Package> = PackageSet::filter_deps(
+                root_id,
+                resolve,
+                has_dev_units,
+                requested_kinds,
+                target_data,
+                force_all_targets,
+            )
+            .iter()
+            .filter_map(|&package_id| {
+                if let Ok(dep_pkg) = self.get_one(package_id) {
+                    if !dep_pkg.targets().iter().any(|t| t.is_lib()) {
+                        Some(dep_pkg)
+                    } else {
+                        None
+                    }
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+            if !pkgs.is_empty() {
+                ret.insert(root_id, pkgs);
+            }
+        });
+
+        ret
+    }
```

Comment:
I think it's possible leverage the `.collect()` to create `BTreeMap` to simplify this to something like
```
fn no_lib_pkgs(...) -> BTreeMap<PackageId, Vec<&Package>> {
  root_ids.iter().map(|root_id| {
      let pkgs = PackageSet::filter_deps(...).iter().filter_map(...);
      (root_id, pkgs)
  }.collect()
}
```

^ this would create a BTreeMap with all root_ids present, and some pointing to empty vecs. Callsite should handle it ok.

Reply:
Addressed. Thanks!